adapter/adapter.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020 Layer5, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adapter
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/client-go/dynamic"
+
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/layer5io/gokit/logger"
+	"github.com/layer5io/meshery-adapter-library/config"
+)
+
+type Handler interface {
+	GetName() string
+	CreateInstance([]byte, string, *chan interface{}) error
+	ApplyOperation(context.Context, OperationRequest) error
+	ListOperations() (Operations, error)
+
+	StreamErr(*Event, error)
+	StreamInfo(*Event)
+}
+
+type BaseHandler struct {
+	Config  config.Handler
+	Log     logger.Handler
+	Channel *chan interface{}
+
+	KubeClient        *kubernetes.Clientset
+	DynamicKubeClient dynamic.Interface
+	KubeConfigPath    string
+	SmiChart          string
+}
+
+type OperationRequest struct {
+	OperationName     string
+	Namespace         string
+	Username          string
+	CustomBody        string
+	IsDeleteOperation bool
+	OperationID       string
+}
+
+func (h *BaseHandler) CreateInstance(kubeconfig []byte, contextName string, ch *chan interface{}) error {
+	h.Channel = ch
+	h.KubeConfigPath = h.Config.GetKey("kube-config-path")
+
+	config, err := h.k8sClientConfig(kubeconfig, contextName)
+	if err != nil {
+		return ErrClientConfig(err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return ErrClientSet(err)
+	}
+
+	h.KubeClient = clientset
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	h.DynamicKubeClient = dynamicClient
+
+	return nil
+}
+
+// creates the namespace unless it is 'default', or it is a delete operation
+func (h *BaseHandler) CreateNamespace(isDelete bool, namespace string) error {
+	if !isDelete && namespace != "default" {
+		if err := h.createNamespace(context.TODO(), namespace); err != nil {
```

Comment:
I saw many of the tools for deploying the manifest to the `apiserver` with `waiting-for-ready`(retry), maybe we can add the  similar type of parameters too.

Reply:
Yes, it would be nice if you could create an issue for that, and/or point to one of those tools?

Reply:
I have created an issue in the `meshery-linkerd` repo, I saw the structure at Helm v3 `pkg/kube` if I remember right, may be we can research more deep with that.

Reply:
Created #6

Reply:
Thank you sir